File: szavazas-core/src/main/java/hu/kdea/szavazas/ballotprocessor/grid/GridDetectionOrchestrator.java
package hu.kdea.szavazas.ballotprocessor.grid;

import boofcv.struct.image.GrayU8;
import hu.kdea.szavazas.ballotprocessor.common.Rect;
import hu.kdea.szavazas.ballotprocessor.debug.GridOverlayDebugRenderer;
import hu.kdea.szavazas.ballotprocessor.debug.ImageSaver;
import hu.kdea.szavazas.ballotprocessor.debug.ProjectionDebugRenderer;
import hu.kdea.szavazas.ballotprocessor.projection.AxisPeaks;
import hu.kdea.szavazas.ballotprocessor.projection.EdgeReconstructor;
import hu.kdea.szavazas.ballotprocessor.projection.PeakFinder;
import hu.kdea.szavazas.ballotprocessor.projection.ProjectionData;
import hu.kdea.szavazas.ballotprocessor.projection.ProjectionUtils;
import java.util.ArrayList;
import java.util.List;
import javax.inject.Inject;
import kotlin.Pair;

public class GridDetectionOrchestrator {
    private final ProjectionDebugRenderer projectionRenderer;
    private final GridOverlayDebugRenderer overlayRenderer;

    @Inject
    public GridDetectionOrchestrator(ImageSaver imageSaver) {
        this.projectionRenderer = imageSaver == null ? null : new ProjectionDebugRenderer(imageSaver);
        this.overlayRenderer = imageSaver == null ? null : new GridOverlayDebugRenderer(imageSaver);
    }

    public List<Rect> detect(GrayU8 binaryClosed, Rect searchRect, int expectedCols, int expectedRows, boolean skipBoundaries, boolean emptySecondColumn) {
        Pair<Rect, ProjectionData> projections = buildProjections(binaryClosed, searchRect, skipBoundaries);
        Rect roi = projections.getFirst();
        ProjectionData data = projections.getSecond();
        int totalCols = emptySecondColumn ? expectedCols + 1 : expectedCols;
        AxisPeaks colAxis = findAxisPeaks(data.getColProj(), data.getColOffset(), totalCols, 0.2, 0.9);
        AxisPeaks rowAxis = findAxisPeaks(data.getRowProj(), data.getRowOffset(), expectedRows, 0.3, 0.7);
        if (projectionRenderer != null) {
            projectionRenderer.renderAll(data, colAxis, rowAxis);
        }
        List<Pair<Integer, Integer>> colEdges = EdgeReconstructor.reconstructViaPairs(data.getColProj(), data.getColOffset(), expectedCols, emptySecondColumn);
        if (colEdges == null) {
            return List.of();
        }
        List<Pair<Integer, Integer>> rowEdges = EdgeReconstructor.reconstructViaPairs(data.getRowProj(), data.getRowOffset(), expectedRows, false);
        if (rowEdges == null) {
            return List.of();
        }
        if (overlayRenderer != null) {
            overlayRenderer.render(binaryClosed, roi, colEdges, rowEdges, data);
        }
        return buildBoxes(rowEdges, colEdges);
    }

    private Pair<Rect, ProjectionData> buildProjections(GrayU8 binaryClosed, Rect searchRect, boolean skipBoundaries) {
        Rect roi = ensureInside(binaryClosed, searchRect);
        Pair<Integer, Integer> boundaries = skipBoundaries
            ? new Pair<>(0, roi.getHeight() - 1)
            : ProjectionUtils.computeRowBoundaries(binaryClosed, roi);
        int cropTop = boundaries.getFirst();
        int cropBottom = boundaries.getSecond();
        int croppedHeight = Math.max(1, cropBottom - cropTop + 1);
        float[] colProjection = ProjectionUtils.columnProjection(binaryClosed, roi, cropTop, cropBottom);
        float[] rowProjection = ProjectionUtils.rowProjection(binaryClosed, roi, cropTop, croppedHeight);
        ProjectionData data = new ProjectionData(colProjection, rowProjection, roi.getX(), roi.getY() + cropTop, roi.getWidth(), croppedHeight);
        return new Pair<>(roi, data);
    }

    private AxisPeaks findAxisPeaks(float[] projection, int offset, int count, double minRatio, double maxRatio) {
        List<Integer> raw = PeakFinder.findRawPeaks(projection, offset);
        List<Integer> merged = PeakFinder.mergeClosePeaks(raw);
        double average = (double) projection.length / count;
        List<Pair<Integer, Integer>> pairs = EdgeReconstructor.pairEdges(merged, (int) (average * minRatio), (int) (average * maxRatio));
        return new AxisPeaks(raw, merged, pairs);
    }

    private List<Rect> buildBoxes(List<Pair<Integer, Integer>> rowEdges, List<Pair<Integer, Integer>> colEdges) {
        List<Rect> boxes = new ArrayList<>();
        for (Pair<Integer, Integer> row : rowEdges) {
            for (Pair<Integer, Integer> col : colEdges) {
                boxes.add(new Rect(col.getFirst(), row.getFirst(), col.getSecond() - col.getFirst(), row.getSecond() - row.getFirst()));
            }
        }
        return boxes;
    }

    private Rect ensureInside(GrayU8 image, Rect rect) {
        int x = Math.max(0, rect.getX());
        int y = Math.max(0, rect.getY());
        return new Rect(x, y, Math.min(rect.getWidth(), image.width - x), Math.min(rect.getHeight(), image.height - y));
    }
}

File: test-utils/src/main/java/hu/kdea/szavazas/ballotprocessor/grid/GridDetectionOrchestrator.kt
package hu.kdea.szavazas.ballotprocessor

import boofcv.struct.image.GrayU8

class GridDetectionOrchestrator(private val debugSaver: ImageSaver? = null) {

    private val projectionRenderer = debugSaver?.let { ProjectionDebugRenderer(it) }
    private val overlayRenderer = debugSaver?.let { GridOverlayDebugRenderer(it) }

    fun detect(
        binaryClosed: GrayU8,
        searchRect: Rect,
        expectedCols: Int,
        expectedRows: Int,
        skipBoundaries: Boolean,
        emptySecondColumn: Boolean
    ): List<Rect> {
        val (roi, data) = buildProjections(binaryClosed, searchRect, skipBoundaries)
        val totalCols = if (emptySecondColumn) expectedCols + 1 else expectedCols
        val colAxis = findAxisPeaks(data.colProj, data.colOffset, totalCols, 0.2, 0.9)
        val rowAxis = findAxisPeaks(data.rowProj, data.rowOffset, expectedRows, 0.3, 0.7)

        logAxis("col", colAxis); logAxis("row", rowAxis)
        projectionRenderer?.renderAll(data, colAxis, rowAxis)

        val colEdges = EdgeReconstructor
            .reconstructViaPairs(data.colProj, data.colOffset, expectedCols, emptySecondColumn)
            ?: return failGrid()
        val rowEdges = EdgeReconstructor
            .reconstructViaPairs(data.rowProj, data.rowOffset, expectedRows, false)
            ?: return failGrid()

        overlayRenderer?.render(binaryClosed, roi, colEdges, rowEdges, data)
        return buildBoxes(rowEdges, colEdges)
    }

    private fun buildProjections(
        binaryClosed: GrayU8, searchRect: Rect, skipBoundaries: Boolean
    ): Pair<Rect, ProjectionData> {
        val roi = ensureInside(binaryClosed, searchRect)
        val (cropTop, cropBottom) =
            if (skipBoundaries) 0 to (roi.height - 1)
            else ProjectionUtils.computeRowBoundaries(binaryClosed, roi)
        val croppedHeight = maxOf(1, cropBottom - cropTop + 1)
        val colProj = ProjectionUtils.columnProjection(binaryClosed, roi, cropTop, cropBottom)
        val rowProj = ProjectionUtils.rowProjection(binaryClosed, roi, cropTop, croppedHeight)
        val data = ProjectionData(colProj, rowProj, roi.x, roi.y + cropTop, roi.width, croppedHeight)
        return roi to data
    }

    private fun findAxisPeaks(
        proj: FloatArray, offset: Int, count: Int, minRatio: Double, maxRatio: Double
    ): AxisPeaks {
        val raw = PeakFinder.findRawPeaks(proj, offset)
        val merged = PeakFinder.mergeClosePeaks(raw)
        val avg = proj.size.toDouble() / count
        val pairs = EdgeReconstructor.pairEdges(merged, (avg * minRatio).toInt(), (avg * maxRatio).toInt())
        return AxisPeaks(raw, merged, pairs)
    }

    private fun logAxis(name: String, axis: AxisPeaks) {
        Logger.d("GridDetector",
            "Raw $name peaks: ${axis.raw}, merged: ${axis.merged}, pairs: ${axis.pairs}")
    }

    private fun failGrid(): List<Rect> {
        Logger.e("GridDetection", "Final grid validation failed")
        return emptyList()
    }

    private fun buildBoxes(
        rowEdges: List<Pair<Int, Int>>, colEdges: List<Pair<Int, Int>>
    ): List<Rect> {
        val boxes = mutableListOf<Rect>()
        for ((top, bottom) in rowEdges)
            for ((left, right) in colEdges)
                boxes.add(Rect(left, top, right - left, bottom - top))
        return boxes
    }

    private fun ensureInside(img: GrayU8, rect: Rect): Rect {
        val x = maxOf(0, rect.x); val y = maxOf(0, rect.y)
        return Rect(x, y, minOf(rect.width, img.width - x), minOf(rect.height, img.height - y))
    }
}